pkg/server/endpoints/endpoints.go

```diff
@@ -289,65 +288,35 @@ func (e *Endpoints) runLocalAccess(ctx context.Context, server *grpc.Server) err
 // getTLSConfig returns a TLS Config hook for the gRPC server
 func (e *Endpoints) getTLSConfig(ctx context.Context) func(*tls.ClientHelloInfo) (*tls.Config, error) {
 	return func(hello *tls.ClientHelloInfo) (*tls.Config, error) {
-		certs, roots, err := e.getCerts(ctx)
-		if err != nil {
-			e.Log.WithError(err).WithField(telemetry.Address, hello.Conn.RemoteAddr().String()).Error("Could not generate TLS config for gRPC client")
-			return nil, err
-		}
-
-		return &tls.Config{
-			// Not all server APIs required a client certificate. Though if one
-			// is presented, verify it.
-			ClientAuth: tls.VerifyClientCertIfGiven,
-
-			Certificates: certs,
-			ClientCAs:    roots,
-
-			MinVersion: tls.VersionTLS12,
-
-			NextProtos: []string{http2.NextProtoTLS},
-		}, nil
+		serverSvid := newX509SVIDSource(e.getServerCertificate)
+		bundle := newBundleSource(func(td *spiffeid.TrustDomain) ([]*x509.Certificate, error) {
+			return e.bundleGetter(ctx, td)
+		})
+
+		spiffeTLSConfig := tlsconfig.MTLSServerConfig(serverSvid, bundle, nil)
+		spiffeTLSConfig.ClientAuth = tls.RequestClientCert
+		spiffeTLSConfig.MinVersion = tls.VersionTLS12
+		spiffeTLSConfig.NextProtos = []string{http2.NextProtoTLS}
+		spiffeTLSConfig.VerifyPeerCertificate = e.serverSpiffeVerificationFunc(bundle)
+
+		return spiffeTLSConfig, nil
 	}
 }
 
-// getCerts queries the datastore and returns a TLS serving certificate(s) plus
-// the current CA root bundle.
-func (e *Endpoints) getCerts(ctx context.Context) ([]tls.Certificate, *x509.CertPool, error) {
-	bundle, err := e.DataStore.FetchBundle(dscache.WithCache(ctx), e.TrustDomain.IDString())
-	if err != nil {
-		return nil, nil, fmt.Errorf("get bundle from datastore: %w", err)
-	}
-	if bundle == nil {
-		return nil, nil, errors.New("bundle not found")
-	}
-
-	var caCerts []*x509.Certificate
-	for _, rootCA := range bundle.RootCas {
-		rootCACerts, err := x509.ParseCertificates(rootCA.DerBytes)
-		if err != nil {
-			return nil, nil, fmt.Errorf("parse bundle: %w", err)
-		}
-		caCerts = append(caCerts, rootCACerts...)
-	}
-
-	caPool := x509.NewCertPool()
-	for _, c := range caCerts {
-		caPool.AddCert(c)
-	}
-
+// getServerCertificate returns the server certificate to be used in the TLS config.
+func (e *Endpoints) getServerCertificate() *tls.Certificate {
 	svidState := e.SVIDObserver.State()
```

Comment:
I wonder if we can change the SVIDObserver to store an x509svid.SVID instead of the tls.Certificate and push the parsing the conversion to the server SVID rotator so we don't have to repeat it on each handshake.